fix(35): update DB Fragmentation and Message Delivery Rate alerts

jkuester opened the pull request:
## Changes

### Alerts

#### DB Fragmentation

After auditing data across a number of instances it is clear that small  Couch DBs tend to have high fragmentation.  CouchDB compaction [only runs for DBs > 1mb](https://docs.couchdb.org/en/stable/config/compaction.html#smoosh.%7Bchannel-name%7D/min_size) (and we are not really concerned if small DBs are fragmented). So, I have updated this alert to only fire if the DB has `< 1000 docs`.  

#### Message Delivery Rate

There was a bug in this alert rule because it was loading data with multiple separate queries, but it could not join the data together when there is multiple CHT instances configured. So, it would work find when there was only one CHT instance, but would just stay in the "No Data" status when more than one CHT instance was being scraped.

I updated the alert rule to just use one big query that can join the multiple sub-queries together by  the `instance` label value.  

#### Client Feedback/Error Rate

The alerting threshold for this was just hard-coded to `> 4`. However, the actual number of feedback docs you might expect for a deployment is not static. Instead it is heavily dependent on how many users you have.  I have updated the alert to have a dynamic threshold that triggers when the feedback rate is higher than 1 doc per 10 users. So, if an instance has 200 users, the alert threshold is `> 20`.  If an instance has 20 users, the threshold is `> 2`.

### Dashboard

On the Overview dashboard, I updated the `DB Conflicts Rate` and `Client Feedback/Error Rate` panels so that their coloring thresholds match the dynamic level from the Client Feedback/Error Rate alert. Sadly, the dynamic colors in the Grafana panels only seem to support binary green/red. So, there is no longer any yellow level on these panels.

## Testing

### DB Fragmentation

The easiest way to test the DB alert changes is to hack up the fake-cht to return some hard-coded data.

- Edit the [`index.js` file](https://github.com/medic/cht-monitoring/blob/main/development/fake-cht/src/index.js) and update the `getCouchDB` function to be:

    ```js
		const getCouchDb = ({ name, update_sequence, doc_count, doc_del_count, fragmentation }) => ({
		  name,
		  update_sequence: update_sequence,
		  doc_count: doc_count,
		  doc_del_count: doc_del_count,
		  fragmentation: fragmentation,
		});
	```

- Edit your [`initial-response.json` file](https://github.com/medic/cht-monitoring/blob/main/development/fake-cht/initial-response.json) to set the desired constant values for the `couchdb` object.  E.g:

	```json
	    "medic": {
	      "name": "medic",
	      "update_sequence": 5350893,
	      "doc_count": 10000,
	      "doc_del_count": 38118,
	      "fragmentation": 10.446299875246408
	    },
	    "sentinel": {
	      "name": "medic-sentinel",
	      "update_sequence": 5252663,
	      "doc_count": 5000,
	      "doc_del_count": 38222,
	      "fragmentation": 3.0206934190548345
	    },
	    "usersmeta": {
	      "name": "medic-users-meta",
	      "update_sequence": 5228371,
	      "doc_count": 2500,
	      "doc_del_count": 38444,
	      "fragmentation": 9.031214446418504
	    },
	    "users": {
	      "name": "_users",
	      "update_sequence": 5340790,
	      "doc_count": 10,
	      "doc_del_count": 38178,
	      "fragmentation": 8.516894610225162
	    }
	```

- Be sure to delete the `development/fake-cht/last-response.json` file if it exists
- Then when you start the fake-cht container, the couchdb metrics should stay constant at the values you set.

The DB Fragmentation alert should be triggered for any db with a fragmentation `> 8` and doc_count `> 1000`.  With the example data from above, `medic`, `usersmeta`, and `users` all have fragmentation `> 8`, but `users` only has `10` docs. So, alerts should only trigger for `medic` and `usersmeta`.

### Message Delivery Rate

To test the Message Delivery Rate, simply configure multiple CHT instances in your `cht-instances.yml` file.  Then view the Alert Rule in the Grafana UI: http://localhost:3000/alerting/grafana/0R-OsCYVz/view and verify that the "State" shown in "Matching Instances" is NOT `NoData`.  (Normal/Alerting/Pending is fine. Just not NoData.)

Good:

![image](https://user-images.githubusercontent.com/8441903/235987545-727909be-3418-4b10-9cc4-414f06396669.png)

Bad: 

![image](https://user-images.githubusercontent.com/8441903/235987657-5aa6f362-e554-40bb-b1c9-fa387c01a05f.png)

### Dashboard Panels

You should be able to test the dashboard panel coloring with some similar hacking of the fake-cht as the `DB Fragmentation` alert test.  Basically update the `index.js` `getConnectedUsers` function to be:

```.js
const getConnectedUsers = ({ count }) => ({
  count,
});
```

Then, in your  [`initial-response.json` file](https://github.com/medic/cht-monitoring/blob/main/development/fake-cht/initial-response.json) you can adjust the `connected_users.count` value.  A _small_ value (e.g. `5`) should result in both  the `DB Conflicts Rate` and `Client Feedback/Error Rate` panels turning red since their threshold will be low with so few users.  A very _large_ value (e.g. 555555555) should result in the panels turning green since the threashold is higher for so many users.

m5r commented on grafana/provisioning/alerting/cht.yml:
```diff
@@ -540,54 +540,44 @@ groups:
+                uid: PBFA97CFB590B2093
+              editorMode: code
+              exemplar: false
+              expr: cht_connected_users_count / 10
```
Neat dynamic threshold!

m5r reviewed (approved):
Nicely done, those dynamic expressions can be very powerful :slightly_smiling_face:

jkuester commented:
Added one more set of changes here. In the dashboard, we had fixed threshold values for setting the color of the panels. I updated the relevant panels to set the color dynamically (with the same value as the alert).  I updated the PR description with more details!

tatilepizs reviewed (approved):
I don't have anything to comment about the code 🙂 
I tested it and everything worked perfectly and smoothly. Thank you for all the instructions and documentation in the description ✅ 

![image](https://user-images.githubusercontent.com/94494491/236509411-db878172-0608-4058-ba11-20add8e4b2f6.png)
